syntax = "proto2";
// Amazon -> UPS
message ACommand {
  repeated AOrderTruck toOrder=1;
  repeated AStartDelivery toStart=2;
  repeated ACheckUsername checkUsers=3;
  repeated int64 acks=4;
}

message AOrderTruck {
  required int64 packageID = 1;
  repeated product productInfo=2;
  required warehouse warehouseInfo=3;
  required destination destinationInfo=4;
  required string upsUsername = 5;
  required int64 seqnum=6;
}

message product{
  required int64 productID=1;
  required string description=2;
  required int32 count = 3;
}

message warehouse{
  required int32 warehouseID = 1;
  required int32 x = 2;
  required int32 y = 3;
}

message destination {
  required int32 x=1;
  required int32 y=2;
}

message AStartDelivery {
  required int64 packageID = 1;
  required int64 seqnum=2;
}

message ACheckUsername {
  required string upsUsername=1;
  required int64 seqnum=2;
}




// UPS -> Amazon
message UCommand {
  repeated UTruckArrival arrived=1;
  repeated UDelivered delivered=2;
  repeated UCheckUsernameResponse checkUser=3;
  repeated UChangeDestination changed=4;
  repeated Err error=5;
  repeated int64 acks=6;
}

// error message
message Err{
  required string err = 1;
  required int64 originseqnum = 2;
  required int64 seqnum = 3;
}

message UChangeDestination {
  required int64 packageID = 1;
  required destination NewDestination=2;
  required int64 seqnum=3;
}

message UTruckArrival {
  required int64 packageID = 1;
  required int32 truckID=2;
  required int64 seqnum=3;
}


message UDelivered {
  required int64 packageID = 1;
  required int64 seqnum=2;
}


message UCheckUsernameResponse {
  required string upsUsername=1;
  //-1: username does not exist
  required int64 upsUserID=2;
  required int64 seqnum=3;
}

